runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/StreamingEngineClient.java

```diff
@@ -0,0 +1,452 @@
+  private final GetWorkRequest getWorkRequest;
+  private final GetWorkBudgetDistributor getWorkBudgetDistributor;
+  private final DispatcherClient dispatcherClient;
+  /**
+   * Reference to {@link GetWorkerMetadataStream} that is lazily initialized, with its initial value
+   * being null.
+   */
+  private final AtomicReference<GetWorkerMetadataStream> getWorkerMetadataStream;
+
+  private final AtomicBoolean isBudgetRefreshPaused;
+  /** Writes are guarded by synchronization, reads are lock free. */
+  private final AtomicReference<StreamEngineConnectionState> connections;
+
+  private final ExecutorService triggeredBudgetRefreshListener;
+  /**
+   * Used to implement publish/subscribe behavior for triggering budget refreshes/redistribution.
+   * Subscriber {@link #triggeredBudgetRefreshListener} waits for messages to be written to the
+   * queue, blocking if necessary.
+   */
+  private final SynchronousQueue<Boolean> budgetRefreshTrigger;
+
+  /** Redistributes the budget on a timed cadence. */
+  private final ScheduledExecutorService scheduledBudgetRefreshExecutor;
+
+  private final ThrottleTimer getWorkerMetadataThrottleTimer;
+  private final CountDownLatch getWorkerMetadataReady;
+  private final ExecutorService consumeWorkerMetadataExecutor;
+
+  private StreamingEngineClient(
+      AtomicBoolean running,
+      GetWorkBudget totalGetWorkBudget,
+      AtomicReference<StreamEngineConnectionState> connections,
+      StreamingEngineStreamFactory streamingEngineStreamFactory,
+      WorkItemReceiver workItemReceiver,
+      WindmillGrpcStubFactory windmillGrpcStubFactory,
+      GetWorkRequest getWorkRequest,
+      GetWorkBudgetDistributor getWorkBudgetDistributor,
+      DispatcherClient dispatcherClient) {
+    this.started = new AtomicBoolean();
+    this.running = running;
+    this.totalGetWorkBudget = totalGetWorkBudget;
+    this.streamingEngineStreamFactory = streamingEngineStreamFactory;
+    this.workItemReceiver = workItemReceiver;
+    this.connections = connections;
+    this.windmillGrpcStubFactory = windmillGrpcStubFactory;
+    this.getWorkRequest = getWorkRequest;
+    this.getWorkBudgetDistributor = getWorkBudgetDistributor;
+    this.dispatcherClient = dispatcherClient;
+    this.isBudgetRefreshPaused = new AtomicBoolean(false);
+    this.budgetRefreshTrigger = new SynchronousQueue<>(/* fair= */ true);
+    this.getWorkerMetadataThrottleTimer = new ThrottleTimer();
+    this.scheduledBudgetRefreshExecutor =
+        createSingleThreadedExecutor(SCHEDULED_BUDGET_REFRESH_THREAD);
+    this.consumeWorkerMetadataExecutor =
+        createSingleThreadedExecutor(CONSUMER_WORKER_METADATA_THREAD);
+    this.triggeredBudgetRefreshListener =
+        createSingleThreadedExecutor(TRIGGERED_BUDGET_REFRESH_THREAD);
+    this.getWorkerMetadataStream = new AtomicReference<>();
+    this.getWorkerMetadataReady = new CountDownLatch(1);
+  }
+
+  public static StreamingEngineClient create(
+      AtomicBoolean running,
```

Comment:
sgtm, i think we would not even need it since the harness class (`StreamingDataflowWorker`), can just not initialize this until it is in a `running` state.

Can make all of these executors use daemon threads then which will get cleaned up and prevent mem leaks.